### PR TITLE
Product Collection: Add stock status filter

### LINKED ISSUE #9362: Product Collection - Filters: Stock Status
### Description
Provide "Stock Status" filter to the Product Collection block. This setting exists in Products block, so should be reused if possible.

### Goals - acceptance criteria
- Changing the filter should impact the query and the change should be reflected in the Editor's preview and frontend
- Filter should be placed in Filters section

### Image
image

### PR DESCRIPTION
**Pull request by imanish003**
This PR adds `Stock status` control to the Product collection block. 
image

Fixes #9362 

### Testing

1. Create a new post
2. Add Product Collection block
3. With the `Product Collection` block selected, check the sidebar for the `Stock status` control. Ensure that it is displayed as shown below:
image

4. Test the functionality of the stock status filter by modifying its selection. Observe that the changes should immediately reflect in the Editor view. For instance, if `On backorder` is the selected option, only products with the `On backorder` status should be visible in the Editor.
5. Save and publish the post.
    - Navigate to the live post on the Frontend and verify that the changes you made in the Editor are accurately reflected.
6. Go back to the Editor and confirm that the filters can be reset using the `Reset` or `Reset all` button, as shown below:
image

* [ ] Do not include in the Testing Notes 

### WooCommerce Visibility

* [ ] WooCommerce Core
* [ ] Feature plugin
* [X] Experimental

### Changelog

> Product Collection: Add stock status control to sidebar settings

### CODE REVIEW

File: src/BlockTypes/ProductCollection.php [hunk 1]
```diff
@@ -326,4 +351,94 @@ private function array_merge_recursive_replace_non_array_properties( $base, $new
...
 		return $base;
 	}
+
+	/**
+	 * Return a query for products depending on their stock status.
+	 *
+	 * @param array $stock_statii An array of acceptable stock statii.
+	 * @return array
+	 */
+	private function get_stock_status_query( $stock_statii ) {
```
Comment (Aljullu) [inline src/BlockTypes/ProductCollection.php]: TIL about _statii_, I guess that's the plural of _status_, no? I think in the codebase we use `statuses` more often than `statii`, so I would rename this variable to `$stock_statuses` if possible to keep it consistent. Not a strong opinion, though.

File: src/BlockTypes/ProductCollection.php [hunk 2]
```diff
@@ -76,46 +76,54 @@ public function update_rest_query( $args, $request ): array {
...
-	 * @return array
+	 * @param array $common_query_values Common query values.
+	 * @param array $query               Query from block context.
 	 */
-	public function extend_rest_query_allowed_params( $params ) {
-		$original_enum             = isset( $params['orderby']['enum'] ) ? $params['orderby']['enum'] : array();
-		$params['orderby']['enum'] = array_unique( array_merge( $original_enum, $this->custom_order_opts ) );
-		return $params;
+	private function get_final_query_args( $common_query_values, $query ) {
+		$orderby_query    = $query['orderby'] ? $this->get_custom_orderby_query( $query['orderby'] ) : [];
```
Comment (Aljullu) [inline src/BlockTypes/ProductCollection.php:100]: This causes a PHP warning in the frontend if `orderby` is not defined:

```
Warning: Undefined array key "orderby" in /wp-content/plugins/woocommerce-blocks/src/BlockTypes/ProductCollection.php on line 100
```

File: src/BlockTypes/ProductCollection.php [hunk 3]
```diff
@@ -326,4 +351,94 @@ private function array_merge_recursive_replace_non_array_properties( $base, $new
...
+		if (
+			count( $stock_statii ) === count( $stock_status_options ) &&
+			array_diff( $stock_statii, $stock_status_options ) === array_diff( $stock_status_options, $stock_statii )
+		) {
+			return array();
+		}
+
+		/**
+		 * If all stock statuses are selected except 'outofstock', we use the
+		 * product visibility query to filter out out of stock products.
```
Comment (Aljullu) [inline src/BlockTypes/ProductCollection.php:381]: I have to say that, as a user, I would find this a bit confusing because the block settings and WC settings can contradict each other. IMO block settings should take priority over WooCommerce settings. So if the _Hide out of stock items from the catalog_ option is checked but I add a Product Collection block selecting to display products with all stock statuses, I would expect all products to appear, ignoring the _Hide out of stock items from the catalog_ option.

In other words, I think the _Hide out of stock items from the catalog_ option should affect which stock status options are selected by default when inserting the block, but they shouldn't affect how the block works once the block attributes have been set.

But... I don't have a strong opinion so I'm happy to change my mind. Also, I don't think it's a blocker.

**Review by Aljullu — APPROVED** [review]
Code looks good and it's working as expected. I left some comments inline, but only the PHP warning is a blocking one IMO, so pre-approving (but feel free to ask for a re-review if you want).

File: src/BlockTypes/ProductCollection.php (hunk 2 above)
Comment (imanish003) [reply src/BlockTypes/ProductCollection.php:100]: Hey @Aljullu, Thanks for raising this 🙏🏻  I have fixed it in [d1e29ef](https://github.com/woocommerce/woocommerce-blocks/pull/9580/commits/d1e29ef02af1acc38614998a11a5a746eef48693) 🤝

File: src/BlockTypes/ProductCollection.php (hunk 1 above)
Comment (imanish003) [reply src/BlockTypes/ProductCollection.php]: Hey @Aljullu, I agree with you. I have updated all instances of `statii` in commit [0c30f5b](https://github.com/woocommerce/woocommerce-blocks/pull/9580/commits/0c30f5b51c4594654ffe5dbb0e937f83f623bde8). Thank you! 🙏🏻

File: src/BlockTypes/ProductCollection.php (hunk 3 above)
Comment (imanish003) [reply src/BlockTypes/ProductCollection.php:381]: Hey @Aljullu,

> I think the Hide out of stock items from the catalog option should affect which stock status options are selected by default when inserting the block, but they shouldn't affect how the block works once the block attributes have been set.

I just want to confirm that currently, when we insert the block, we do take the "Hide out of stock items from the catalog" setting into consideration. This is implemented [here](https://github.com/woocommerce/woocommerce-blocks/blob/0c30f5b51c4594654ffe5dbb0e937f83f623bde8/assets/js/blocks/product-collection/constants.ts#L18-L32).

Also, I agree that it makes sense to ignore the "Hide out of stock items from the catalog" setting if we allow merchants to explicitly set stock statuses in the Product Collection block.
image


PS: I will go ahead and merge this PR for now. We can continue our discussion here. If we decide to ignore the global "Hide out of stock items from the catalog" setting, I will create a new PR for that.

CC: @nerrad @tjcafferkey

Comment (Aljullu) [reply src/BlockTypes/ProductCollection.php:381]: > I just want to confirm that currently, when we insert the block, we do take the "Hide out of stock items from the catalog" setting into consideration. This is implemented [here](https://github.com/woocommerce/woocommerce-blocks/blob/0c30f5b51c4594654ffe5dbb0e937f83f623bde8/assets/js/blocks/product-collection/constants.ts#L18-L32).

Oh right, I missed it! That's nice.
